File: curriculum/01_rate_limiter/rust-impl/src/middleware.rs
//! axum middleware that enforces the token bucket on the rate-limited route.
//!
//! The middleware:
//!  1. Extracts the client IP from the [`ConnectInfo`] (provided by
//!     `into_make_service_with_connect_info` at serve time).
//!  2. Asks the [`RateLimiter`] for a [`Decision`].
//!  3. On allow: calls the inner handler, then stamps the three
//!     `X-RateLimit-*` headers on the response.
//!  4. On deny: short-circuits with `429`, a JSON body, and
//!     `X-RateLimit-*` + `Retry-After` headers.

use std::net::SocketAddr;
use std::sync::Arc;

use axum::body::Body;
use axum::extract::{ConnectInfo, State};
use axum::http::{HeaderName, HeaderValue, Request, StatusCode};
use axum::middleware::Next;
use axum::response::{IntoResponse, Response};
use axum::Json;
use serde_json::json;

use crate::rate_limiter::{Decision, RateLimiter};

const HDR_LIMIT: HeaderName = HeaderName::from_static("x-ratelimit-limit");
const HDR_REMAINING: HeaderName = HeaderName::from_static("x-ratelimit-remaining");
const HDR_RESET: HeaderName = HeaderName::from_static("x-ratelimit-reset");
const HDR_RETRY_AFTER: HeaderName = HeaderName::from_static("retry-after");

/// axum middleware fn. Use with [`axum::middleware::from_fn_with_state`].
pub async fn rate_limit_middleware(
    State(limiter): State<Arc<RateLimiter>>,
    ConnectInfo(addr): ConnectInfo<SocketAddr>,
    req: Request<Body>,
    next: Next,
) -> Response {
    let ip = addr.ip();
    match limiter.check(ip) {
        Decision::Allowed {
            remaining,
            limit,
            reset_epoch,
        } => {
            let mut response = next.run(req).await;
            stamp_headers(response.headers_mut(), limit, remaining, reset_epoch, None);
            response
        }
        Decision::Denied {
            remaining,
            limit,
            reset_epoch,
            retry_after,
        } => {
            let body = Json(json!({
                "error": "Too Many Requests",
                "retry_after_seconds": retry_after,
            }));
            let mut response = (StatusCode::TOO_MANY_REQUESTS, body).into_response();
            stamp_headers(
                response.headers_mut(),
                limit,
                remaining,
                reset_epoch,
                Some(retry_after),
            );
            response
        }
    }
}

fn stamp_headers(
    headers: &mut axum::http::HeaderMap,
    limit: u64,
    remaining: u64,
    reset_epoch: u64,
    retry_after: Option<u64>,
) {
    headers.insert(HDR_LIMIT, HeaderValue::from(limit));
    headers.insert(HDR_REMAINING, HeaderValue::from(remaining));
    headers.insert(HDR_RESET, HeaderValue::from(reset_epoch));
    if let Some(ra) = retry_after {
        headers.insert(HDR_RETRY_AFTER, HeaderValue::from(ra));
    }
}

File: curriculum/01_rate_limiter/go-impl/ratelimit/middleware.go
package ratelimit

import (
	"encoding/json"
	"net"
	"net/http"
	"strconv"
)

// ClientKey extracts the client identifier used as the bucket key. We use the
// TCP peer address (the closest thing to a "client_ip" the spec asks for) and
// strip the port. In deployments behind a trusted reverse proxy you would
// also consult X-Forwarded-For, but the spec doesn't require it and trusting
// unvalidated forwarded headers is a known footgun.
func ClientKey(r *http.Request) string {
	host, _, err := net.SplitHostPort(r.RemoteAddr)
	if err != nil {
		// RemoteAddr might be a bare host (e.g. in some test harnesses).
		return r.RemoteAddr
	}
	return host
}

// Middleware returns an http.Handler middleware that enforces the limiter on
// every incoming request. If the bucket has a token, the wrapped handler is
// invoked and rate-limit headers are set. If not, a 429 is returned with
// Retry-After and a JSON body.
func (rl *RateLimiter) Middleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		key := ClientKey(r)
		d := rl.Allow(key)

		h := w.Header()
		h.Set("X-RateLimit-Limit", strconv.Itoa(d.Limit))
		h.Set("X-RateLimit-Remaining", strconv.Itoa(d.Remaining))
		h.Set("X-RateLimit-Reset", strconv.FormatInt(d.Reset, 10))

		if !d.Allowed {
			h.Set("Retry-After", strconv.Itoa(d.RetryAfter))
			h.Set("Content-Type", "application/json")
			// Use WriteHeader explicitly so json.Encoder writes only the
			// body, not an extra status line.
			w.WriteHeader(http.StatusTooManyRequests)
			_ = json.NewEncoder(w).Encode(rateLimitResponse{
				Error:             "Too Many Requests",
				RetryAfterSeconds: d.RetryAfter,
			})
			return
		}
		next.ServeHTTP(w, r)
	})
}

// StatusHandler returns an http.Handler that reports the caller's current
// bucket state. It is intentionally NOT rate-limited — the spec says so
// explicitly, and rate-limiting your observability endpoint is a common
// production mistake that breaks debugging.
func (rl *RateLimiter) StatusHandler() http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		key := ClientKey(r)
		tokens, _ := rl.Snapshot(key)

		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(statusResponse{
			ClientIP:            key,
			TokensRemaining:     tokens,
			MaxCapacity:         rl.capacity,
			RefillRatePerSecond: rl.refillPerSec,
		})
	})
}
